File: packages/media_library/lib/src/media_record.dart
// ============================================================
// As Above, So Below. As Within, So Without.
// The Future Dictates the Past and the Past is Always Present.
// ============================================================
import 'package:metadata_core/metadata_core.dart';

import 'media_kind.dart';
import 'media_source.dart';
import 'thumb.dart';

class MediaRecord {
  const MediaRecord({
    required this.source,
    required this.kind,
    required this.status,
    required this.sortIndex,
    this.photo,
    this.video,
    this.warnings = const <String>[],
    this.thumb = const ThumbInfo(ThumbStatus.pending),
  });

  final MediaSource source;
  final MediaKind kind;
  final ExtractStatus status;
  final int sortIndex;
  final PhotoMeta? photo;
  final VideoMeta? video;
  final List<String> warnings;
  final ThumbInfo thumb;

  factory MediaRecord.fromPhoto({
    required MediaSource source,
    required int sortIndex,
    required PhotoMeta photo,
    required ExtractStatus status,
    List<String> warnings = const <String>[],
  }) {
    return MediaRecord(
      source: source,
      kind: MediaKind.photo,
      status: status,
      sortIndex: sortIndex,
      photo: photo,
      warnings: List<String>.unmodifiable(warnings),
    );
  }

  factory MediaRecord.fromVideo({
    required MediaSource source,
    required int sortIndex,
    required VideoMeta video,
    required ExtractStatus status,
    List<String> warnings = const <String>[],
  }) {
    return MediaRecord(
      source: source,
      kind: MediaKind.video,
      status: status,
      sortIndex: sortIndex,
      video: video,
      warnings: List<String>.unmodifiable(warnings),
    );
  }

  factory MediaRecord.unsupported({
    required MediaSource source,
    required int sortIndex,
    List<String> warnings = const <String>['unsupported type'],
  }) {
    return MediaRecord(
      source: source,
      kind: MediaKind.unsupported,
      status: ExtractStatus.unsupported,
      sortIndex: sortIndex,
      warnings: List<String>.unmodifiable(warnings),
      thumb: const ThumbInfo(
        ThumbStatus.unavailable,
        reason: 'unsupported type',
      ),
    );
  }

  Map<String, dynamic> toJson() => <String, dynamic>{
        'source': source.toJson(),
        'kind': kind.name,
        'status': status.name,
        'sortIndex': sortIndex,
        'photo': photo == null ? null : photoMetaToJson(photo!),
        'video': video == null ? null : videoMetaToJson(video!),
        'warnings': warnings,
        'thumb': thumb.toJson(),
      };

  static MediaRecord fromJson(Map<String, dynamic> json) {
    MediaKind kind = MediaKind.unsupported;
    for (final MediaKind candidate in MediaKind.values) {
      if (candidate.name == json['kind']) {
        kind = candidate;
      }
    }
    return MediaRecord(
      source: MediaSource.fromJson(
        (json['source'] as Map).cast<String, dynamic>(),
      ),
      kind: kind,
      status: _parseStatus(json['status']),
      sortIndex: json['sortIndex'] as int,
      photo: json['photo'] == null
          ? null
          : photoMetaFromJson(
              (json['photo'] as Map).cast<String, dynamic>(),
            ),
      video: json['video'] == null
          ? null
          : videoMetaFromJson(
              (json['video'] as Map).cast<String, dynamic>(),
            ),
      warnings: (json['warnings'] as List).cast<String>(),
      thumb: ThumbInfo.fromJson(
        (json['thumb'] as Map).cast<String, dynamic>(),
      ),
    );
  }

  static ExtractStatus _parseStatus(Object? raw) {
    for (final ExtractStatus candidate in ExtractStatus.values) {
      if (candidate.name == raw) {
        return candidate;
      }
    }
    return ExtractStatus.partial;
  }
}

File: native/vision-core/src/main/kotlin/com/visionengine/core/MediaRecord.kt
// ============================================================
// As Above, So Below. As Within, So Without.
// The Future Dictates the Past and the Past is Always Present.
// ============================================================
package com.visionengine.core

data class MediaRecord(
    val source: MediaSource,
    val kind: MediaKind,
    val status: ExtractStatus,
    val sortIndex: Int,
    val photo: PhotoMeta? = null,
    val video: VideoMeta? = null,
    val warnings: List<String> = emptyList(),
    val thumb: ThumbInfo = ThumbInfo(ThumbStatus.pending),
) {
    fun toMap(): Map<String, Any?> = mapOf(
        "source" to source.toMap(),
        "kind" to kind.name,
        "status" to status.name,
        "sortIndex" to sortIndex,
        "photo" to photo?.let(::photoMetaToMap),
        "video" to video?.let(::videoMetaToMap),
        "warnings" to warnings,
        "thumb" to thumb.toMap(),
    )

    companion object {
        fun fromPhoto(
            source: MediaSource,
            sortIndex: Int,
            photo: PhotoMeta,
            status: ExtractStatus,
            warnings: List<String> = emptyList(),
        ): MediaRecord = MediaRecord(
            source = source,
            kind = MediaKind.photo,
            status = status,
            sortIndex = sortIndex,
            photo = photo,
            warnings = warnings.toList(),
        )

        fun fromVideo(
            source: MediaSource,
            sortIndex: Int,
            video: VideoMeta,
            status: ExtractStatus,
            warnings: List<String> = emptyList(),
        ): MediaRecord = MediaRecord(
            source = source,
            kind = MediaKind.video,
            status = status,
            sortIndex = sortIndex,
            video = video,
            warnings = warnings.toList(),
        )

        fun unsupported(
            source: MediaSource,
            sortIndex: Int,
            warnings: List<String> = listOf("unsupported type"),
        ): MediaRecord = MediaRecord(
            source = source,
            kind = MediaKind.unsupported,
            status = ExtractStatus.unsupported,
            sortIndex = sortIndex,
            warnings = warnings.toList(),
            thumb = ThumbInfo(
                ThumbStatus.unavailable,
                reason = "unsupported type",
            ),
        )

        @Suppress("UNCHECKED_CAST")
        fun fromMap(json: Map<String, Any?>): MediaRecord {
            var kind = MediaKind.unsupported
            for (candidate in MediaKind.values()) {
                if (candidate.name == json["kind"]) {
                    kind = candidate
                }
            }
            val photoNode = json["photo"] as Map<String, Any?>?
            val videoNode = json["video"] as Map<String, Any?>?
            val warningNodes = json["warnings"] as List<*>
            val thumbNode = json["thumb"] as Map<String, Any?>
            return MediaRecord(
                source = MediaSource.fromMap(
                    json["source"] as Map<String, Any?>,
                ),
                kind = kind,
                status = parseStatus(json["status"]),
                sortIndex = (json["sortIndex"] as Number).toInt(),
                photo = photoNode?.let(::photoMetaFromMap),
                video = videoNode?.let(::videoMetaFromMap),
                warnings = warningNodes.filterIsInstance<String>(),
                thumb = ThumbInfo.fromMap(thumbNode),
            )
        }

        private fun parseStatus(raw: Any?): ExtractStatus {
            for (candidate in ExtractStatus.values()) {
                if (candidate.name == raw) {
                    return candidate
                }
            }
            return ExtractStatus.partial
        }
    }
}
